[WIP] 576 Add methods to check SSH connections/commands with retry

ztbrown opened the pull request:
Early PR for feedback on approach for addressing #576 

@rhoboat, I'm working on the helper methods you suggested in 576 and I want to make sure this is an approach you are comfortable with. 

I wrote `CheckSshConnectionWithRetry` and `CheckSshConnectionWithRetryE` to accept an optional function. They default to wrapping `CheckSshConnectionE`, but the option to inject means we can swap it for a mock function and unit test much easier. My worry is that the interface might not be as clear as it could be to users. What do you think? 

If this looks good, I'll move on to `CheckSshCommandWithRetry` and `CheckSshCommandWithRetryE`. Thanks!

brikis98 commented:
Just a heads up that @rhoboat is currently OOO, but they will take a look when they are back!

ztbrown commented:
hey @rhoboat! Bumping this for visibility. I can keep going, make alterations or close the PR. Thanks!

ztbrown commented:
Bump.

brikis98 commented:
@rhoboat Could you take a look?

@ztbrown Could you pull in the latest from `master`?

ztbrown commented:
rebased against master. This is still what I would consider a WIP or early PR, looking for feedback. Do we like this approach? Is it too clever by half? If we don't like it, how would you like to proceed? 

Thanks!

ztbrown commented:
Bump. If you all would like me to drop it, I will close. LMK!

rhoboat commented:
This looks good! I'm taking a closer look now...

rhoboat commented:
I think you're totally on the right track @ztbrown. Keep going, and feel free to ping me again when you're ready for a final review.

ztbrown commented:
That rocks! Thank you. I'll set aside some time to jump back in. Thanks again!

ztbrown commented:
Thank you but it's totally understandable. :)

ztbrown commented:
OK @rhoboat, looking for a final review. I've added the retry logic to `CheckSshCommandWithRetry` and `CheckSshCommandWithRetryE`. Added some tests for when retries get exceeded on `CheckSshCommandWithRetryE` and `CheckSshConnectionWithRetryE`. Let me know what you think. Thanks!

rhoboat commented:
It's odd that the checks for this PR are not running, i.e., tests. I'd like to see them succeed on CI.

infraredgirl commented:
I just kicked off the tests.

ztbrown commented:
Looks like an unrelated failure. Do you typically re-kick the tests in this situation?

ztbrown commented:
lmk if you want me to rebase against master, squash down to one commit or just press the "resolve conflicts" button above.

yorinasub17 commented:
@ztbrown Would be best if you can rebase. Since it's a `go.sum` conflict, this is typically better to resolve by wiping out the file and running `go mod tidy` to ensure all the dependencies resolve correctly.

ztbrown commented:
awesome. I'm on it.

yorinasub17 commented:
Thanks! Just kicked off another build.

yorinasub17 reviewed (approved):
LGTM! And the test failure is unrelated to this change, so will move forward with merging this in! Thanks for your patience getting this over the finish line!

ztbrown commented:
thank you everyone! 🎉
